### PR TITLE
Added Support for guided decoding in offline interface

### PR DESCRIPTION
**Pull request by kevinbu233**
Added support for guided decoding (JSON, regex, choice) on offline interface that is raised in issue #3536 . 
This is a continuation and build off the work of #2815 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/utils.py [hunk 1]
```diff
@@ -51,7 +52,7 @@ def reset(self) -> None:
 class LRUCache(Generic[T]):
 
     def __init__(self, capacity: int):
-        self.cache = OrderedDict[Hashable, T]()
+        self.cache: OrderedDict[Hashable, T] = OrderedDict()
```
Comment (simon-mo) [inline vllm/utils.py:55]: I'm surprised this works...

File: vllm/sampling_params.py [hunk 2]
```diff
@@ -188,6 +203,8 @@ def __init__(
         # injected by the engine
         self.eos_token_id = None
 
+    
+
```
Comment (simon-mo) [inline vllm/sampling_params.py:207]:
```suggestion
```

**Review by simon-mo — DISMISSED** [review]
please fix merge conflict and run formatting script.

File: tests/entrypoints/test_local_LLM.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+start: select_statement
+
+select_statement: "SELECT" column "from" table "where" condition
+
+column: "col_1" | "col_2"
+table: "table_1" | "table_2"
+condition: column "=" number
+
+number: "1" | "2"
+"""
```
Comment (simon-mo) [inline tests/entrypoints/test_local_LLM.py:137]: Move this to fixture as well

File: vllm/model_executor/guided_decoding/outlines_decoding.py [hunk 4]
```diff
@@ -82,9 +82,78 @@ async def get_outlines_guided_decoding_logits_processor(
...
+
+    guide, mode = _get_guide_and_mode_from_sampling_params(
+        sampling_params.guided_options)
+    if not guide:
+        return None
+
+    # if global_thread_pool is None:
+    #     global_thread_pool = concurrent.futures.ThreadPoolExecutor(
+    #         max_workers=2)
+
```
Comment (simon-mo) [inline vllm/model_executor/guided_decoding/outlines_decoding.py:101]: remove

File: vllm/model_executor/guided_decoding/outlines_decoding.py [hunk 5]
```diff
@@ -82,9 +82,78 @@ async def get_outlines_guided_decoding_logits_processor(
...
+    elif "guided_grammar" in guided_options:
+        return guided_options["guided_grammar"], GuidedDecodingMode.GRAMMAR
+    else:
+        return None, None
+
+
 def _get_guide_and_mode(
     request: Union[CompletionRequest, ChatCompletionRequest]
-) -> Union[Tuple[str, GuidedDecodingMode], Tuple[None, None]]:
+) -> Tuple[str, GuidedDecodingMode]:
```
Comment (simon-mo) [inline vllm/model_executor/guided_decoding/outlines_decoding.py:156]: not sure this is correct because it can still return `(None, None)`

File: vllm/sampling_params.py [hunk 6]
```diff
@@ -129,6 +129,7 @@ def __init__(
         spaces_between_special_tokens: bool = True,
         logits_processors: Optional[List[LogitsProcessor]] = None,
         truncate_prompt_tokens: Optional[Annotated[int, Field(ge=1)]] = None,
+        guided_options: Optional[Dict[str, Union[list, str, dict]]] = None
```
Comment (simon-mo) [inline vllm/sampling_params.py:132]: Please add docstring (see other options) in the class docstring above.

File: tests/entrypoints/test_local_LLM.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    "Hello, my name is",
+    "The president of the United States is",
+    "The capital of France is",
+    "The future of AI is",
+]
+
+
+@pytest.fixture(scope="session")
+def llm():
+    return LLM(model=MODEL_NAME, max_model_len=15000)
```
Comment (jackboyla) [inline tests/entrypoints/test_local_LLM.py:24]:
```suggestion
    return LLM(model=MODEL_NAME, max_model_len=128)
```

**Comment by simon-mo** [comment]
@br3no can you help review this? (also @rkooo567 if you have bandwidth) 🙏

File: vllm/entrypoints/llm.py [hunk 8]
```diff
@@ -493,24 +495,51 @@ def _validate_and_add_requests(
...
-        # Add requests to the engine.
         for i, request_inputs in enumerate(inputs):
             self._add_request(
                 request_inputs,
                 params[i] if isinstance(params, Sequence) else params,
                 lora_request=lora_request,
             )
 
+    def _add_guided_processor(self, params: SamplingParams):
+        if options := params.guided_options:
```
Comment (br3no) [inline vllm/entrypoints/llm.py:528]: **Suggestion**

I would suggest checking for `None` instead. Makes the code more accessible to everyone working on this project.

File: vllm/model_executor/guided_decoding/__init__.py [hunk 9]
```diff
@@ -1,25 +1,47 @@
...
+    if request.guided_decoding_backend == 'outlines':
+        return get_outlines_guided_decoding_logits_processor(
             request, tokenizer)
-    if guided_decoding_backend == 'lm-format-enforcer':
-        return await get_lm_format_enforcer_guided_decoding_logits_processor(
+    if request.guided_decoding_backend == 'lm-format-enforcer':
+        ## Import moved inside function to avoide circular
+        ## import with vllm.entrypoints.LLM.py
+        from vllm.model_executor.guided_decoding.lm_format_enforcer_decoding import (  # noqa
+            get_lm_format_enforcer_guided_decoding_logits_processor)
```
Comment (br3no) [inline vllm/model_executor/guided_decoding/__init__.py:38]: **Question**

Sure this is necessary?

File: vllm/model_executor/guided_decoding/outlines_decoding.py [hunk 10]
```diff
@@ -53,38 +50,45 @@ class GuidedDecodingMode(Enum):
...
-    logits_processor = copy(result)
+    logits_processor = copy(
+        _get_cached_logits_processor(guide, tokenizer, mode,
+                                     request.guided_whitespace_pattern))
     # reset logits processor's internal state
     logits_processor.init_state()
     return logits_processor
 
 
+def convert_json_format(json):
```
Comment (br3no) [inline vllm/model_executor/guided_decoding/outlines_decoding.py:74]: **Question**

Where is this used?

File: vllm/model_executor/guided_decoding/outlines_decoding.py [hunk 11]
```diff
@@ -53,38 +50,45 @@ class GuidedDecodingMode(Enum):
...
+        # turn dict into hashable string
+        json = json_dumps(json)
+    elif isinstance(json, BaseModel):
+        # use pydantic signature so that different model classes
+        # with the same fields will get hashed the same
+        json = str(json.__signature__)
+    return json
+
+
+def convert_guided_choice_format(guided_choice):
```
Comment (br3no) [inline vllm/model_executor/guided_decoding/outlines_decoding.py:85]: **Question**

Where is this used?

**Review by br3no — CHANGES_REQUESTED** [review]
Maybe it would be easier to split the PR into one refactoring PR and one PR for adding the new feature of supporting guided decoding in offline inference?

File: vllm/model_executor/guided_decoding/__init__.py [hunk 12]
```diff
@@ -1,25 +1,47 @@
...
+
+    return await loop.run_in_executor(
+        global_thread_pool,
+        get_guided_decoding_logits_processor,
+        request,
+        tokenizer,
+    )
+
+
+def get_guided_decoding_logits_processor(
```
Comment (maxdebayser) [inline vllm/model_executor/guided_decoding/__init__.py:29]: For the offline use case it makes sense to have a non-async way to get the processor. But orthogonal to that is the issue of sharing of stateful logits processors. In PR #5329 both concerns are addressed because it adds a factory with an async and a sync builder method.

File: vllm/sampling_params.py [hunk 13]
```diff
@@ -104,6 +107,8 @@ class SamplingParams:
         truncate_prompt_tokens: If set to an integer k, will use only the last k
             tokens from the prompt (i.e., left truncation). Defaults to None
             (i.e., no truncation).
+        guided_options: Configuration dictionary for guided decoding. Refer to
+            the `GuidedDecodingFields` class for the available options.
```
[the remark this answers is not in the document]
Comment (maxdebayser) [reply vllm/sampling_params.py:111]: Yes, if logits processors and guided options are both in the sampling params options then you have both the object and the arguments necessary to build it. I agree that this is somewhat redundant.

**Comment by DarkLight1337** [comment]
Closing as superseded by #6878.
